circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/recursion/honk/transcript/trancript.hpp

```diff
@@ -0,0 +1,174 @@
+
+    /**
+     * @brief Compute the challenges (more than 1) indicated by labels
+     *
+     * @tparam Strings
+     * @param labels Names of the challenges to be computed
+     * @return std::array<FF, sizeof...(Strings)> Array of challenges
+     */
+    template <typename... Strings> std::array<field_pt, sizeof...(Strings)> get_challenges(const Strings&... labels)
+    {
+        // Compute the indicated challenges from the native transcript
+        constexpr size_t num_challenges = sizeof...(Strings);
+        std::array<FF, num_challenges> native_challenges{};
+        native_challenges = native_transcript.get_challenges(labels...);
+
+        /*
+         * TODO(luke): Do stdlib hashing here. E.g., for the current pedersen/blake setup, we could write data into a
+         * byte_array as it is received from prover, then compress via pedersen and apply blake3s. Not doing this now
+         * since it's a pain and we'll be revamping our hashing anyway. For now, simply convert the native hashes to
+         * stdlib types without adding any hashing constraints.
+         */
+        std::array<field_pt, num_challenges> challenges;
+        for (size_t i = 0; i < num_challenges; ++i) {
+            challenges[i] = native_challenges[i];
+        }
+
+        return challenges;
+    }
+
+    /**
+     * @brief Compute the single challenge indicated by the input label
+     *
+     * @param label Name of challenge
+     * @return field_pt Challenge
+     */
+    field_pt get_challenge(const std::string& label)
+    {
+        // Compute the indicated challenge from the native transcript
+        auto native_challenge = native_transcript.get_challenge(label);
+
+        // TODO(luke): Stdlib hashing here...
+
+        return field_pt(native_challenge);
+    }
+
+    /**
+     * @brief Extract a native element from the transcript and return a corresponding stdlib type
+     *
+     * @tparam T Type of the native element to be extracted
+     * @param label Name of the element
+     * @return The corresponding element of appropriate stdlib type
+     */
+    template <class T> auto receive_from_prover(const std::string& label)
+    {
+        // Extract the native element from the native transcript
+        T element = native_transcript.template receive_from_prover<T>(label);
+
+        // Add variables corresponding to the witness element and return the corresponding stdlib type
+        return stdlib_type_from_witness(element);
+    }
+
+  private:
+    // Series of overloaded methods for converting native types extracted from the native transcript to the
+    // corresponding stdlib type for output.
+    // TODO(luke): Eventually these can also add data to a buffer (byte_array) to be hashed.
+
+    /**
+     * @brief Construct stdlib field from uint32_t
+     *
+     * @param element
+     * @return field_pt
+     */
+    field_pt stdlib_type_from_witness(uint32_t native_element)
```

Comment:
Could you use existing functionality? for this and similar functions below? For instance, we have a function `field_t<Builder>::from_witness`. These things are generally useful and therefore should be constructors or member functions on the appropriate types.

Reply:
I'm not sure I follow. Do you just mean e.g. change this from `witness_pt(builder, native_element);` to `field_pt::from_witness(builder, native_element)`?

Reply:
I mean rather than defining these little utility functions in the class, you should use existing functions or put the utility functions in the appropriate classes -- if you're interested in having these constructor-like functions, then somebody else might be. If you're hesitant to put them there then maybe they shouldn't exist at all?

Reply:
I'm having a hard time finding a clean solution along the lines of what you're describing. I've taken a half-measure for now and moved the "from witness" functions into a utility class in the stdlib so at least they are separated from the Transcript. 

It's possible I'm not understanding your suggestion well but an example of why it seems difficult: The template type for `receive_from_prover()` is currently the native type to be extracted and there are no existing methods that map native types to stdlib types, as far as I know. In theory I could add a function `to_field_t()` or something to `field`, but as far as I can tell the native types don't currently know anything about stdlib types so that feels like a major departure.  Then there are types like uint32_t (which I cant add methods to unless I create a wrapper for the c++ std library type) and std::array<fr> which is not a first class type to which I could add methods.

We do have `from_witness` methods implemented on the stdlib field/group types but I still need to connect the stdlib type to the native type that I'm trying to extract from the native transcript. So for example if I changed the template parameter in `receive_from_prover()` to a stdlib type, I'm not sure how I would distinguish between extracting a uint32_t or fr from the native transcript.